agent/api/task/task.go

```diff
@@ -953,6 +1014,91 @@ func (task *Task) generateENIExtraHosts() []string {
 	return extraHosts
 }
 
+// shouldOverridePIDMode returns true if the PIDMode of the container needs
+// to be overridden. It also returns the override string in this case. It returns
+// false otherwise
+func (task *Task) shouldOverridePIDMode(container *apicontainer.Container, dockerContainerMap map[string]*apicontainer.DockerContainer) (bool, string) {
+	// If the container is an internal container (ContainerEmptyHostVolume,
+	// ContainerCNIPause, or ContainerNamespacePause), then PID namespace for
+	// the container itself should be private (default Docker option)
+	if container.IsInternal() {
+		return false, ""
+	}
+
+	if task.getPIDMode() == pidModeHost {
+		return true, pidModeHost
+	} else if task.getPIDMode() == pidModeTask {
+		pauseContName, ok := task.findPauseContainerName(apicontainer.ContainerNamespacePause)
+		if !ok {
+			seelog.Critical("Namespace Pause container not found in the task: %s", task.String())
+			return false, ""
+		}
+		pauseToGiveDocker, ok := dockerContainerMap[pauseContName]
+		if !ok || pauseToGiveDocker == nil {
+			// Docker container shouldn't be nill or not exist if the Container definition within task exists; implies code-bug
+			seelog.Criticalf("Namespace Pause container required, but not found in container map for container: [%s] in task: %s",
```

Comment:
This is a failure case, which should be handled in a way that the task is not run. If pidMode is task and pause container is not found, we should stop the task imo. same for ipcMode.